test/composableStableJoin.integration.test.ts

```diff
@@ -0,0 +1,309 @@
+
+    beforeAll(async () => {
+        // setup mock api
+        const api = new MockApi();
+
+        // get pool state from api
+        const poolInput = await api.getPool(poolId);
+
+        const client = createTestClient({
+            mode: 'anvil',
+            chain: CHAINS[chainId],
+            transport: http(rpcUrl),
+        })
+            .extend(publicActions)
+            .extend(walletActions);
+
+        txInput = {
+            client,
+            poolJoin: new PoolJoin(),
+            slippage: Slippage.fromPercentage('1'), // 1%
+            poolInput,
+            testAddress: '0x10a19e7ee7d7f8a52822f6817de8ea18204f2e4f', // Balancer DAO Multisig
+            joinInput: {} as JoinInput,
+            checkNativeBalance: false,
+            chainId,
+        };
+
+        // setup BPT token
+        bptToken = new Token(chainId, poolInput.address, 18, 'BPT');
+    });
+
+    beforeEach(async () => {
+        await forkSetup(
+            txInput.client,
+            txInput.testAddress,
+            [...txInput.poolInput.tokens.map((t) => t.address)],
+            [0, 0, 3],
+            [
+                ...txInput.poolInput.tokens.map((t) =>
+                    parseUnits('100', t.decimals),
+                ),
+            ]
+        );
+    });
+
+    test('unbalanced join', async () => {
+        const bptIndex = txInput.poolInput.tokens.findIndex(
+            (t) => t.address === txInput.poolInput.address,
+        );
+        const poolTokensWithoutBpt = txInput.poolInput.tokens
+            .map((t) => new Token(chainId, t.address, t.decimals))
+            .filter((_, index) => index !== bptIndex);
+
+        const amountsIn = poolTokensWithoutBpt.map((t) =>
+            TokenAmount.fromHumanAmount(t, '1'),
+        );
+
+        // perform join query to get expected bpt out
+        const joinInput: UnbalancedJoinInput = {
+            amountsIn,
+            chainId,
+            rpcUrl,
+            kind: JoinKind.Unbalanced,
+        };
+
+        const { queryResult, maxAmountsIn, minBptOut, value } = await doJoin({
+            ...txInput,
+            joinInput,
+        });
+        // Query should use same amountsIn as user sets
+        expect(
+            queryResult.amountsIn.filter((_, index) => index !== bptIndex),
+        ).to.deep.eq(amountsIn);
+        expect(queryResult.tokenInIndex).to.be.undefined;
+
+        // Should be no native value
+        expect(value).to.be.undefined
+
+        // Expect some bpt amount
+        expect(queryResult.bptOut.amount > BigInt(0)).to.be.true;
+
+        // Confirm slippage - only bpt out
+        const expectedMinBpt = txInput.slippage.removeFrom(
+            queryResult.bptOut.amount,
+        );
+        expect(expectedMinBpt).to.deep.eq(minBptOut);
+        const expectedMaxAmountsIn = amountsIn.map((a) => a.amount);
+        expect(expectedMaxAmountsIn).to.deep.eq(
+            maxAmountsIn.filter((_, index) => index !== bptIndex),
+        );
```

Comment:
I see that we have a few test assertions within `doJoin` and some other spread between tests. How similar are they between tests? Does it make sense to extract all checks to a separate `assertResults` function?
This way we could separate join functionality (`doJoin`) and test assertions (`assertResults`).

Reply:
Yeah, maybe I can move these assertions back to the test file rather than in a helper function, wdyt?

Reply:
Yeah, not only having them all within the test files, but reusing them would be great. A helper function within the test file seems to be a good solution to achieve that.

Reply:
@brunoguerios - I've done a refactor of the tests, take a look and let me know what you think.